TESTS/mbed_drivers/timerevent/main.cpp

```diff
@@ -0,0 +1,239 @@
+ * Given an instance of @a TimerEvent subclass
+ * When a tiestamp is set with @a insert_absolute()
+ *     and timestamp is removed before being reached
+ * Then the event handler is never called
+ */
+template<typename T>
+void test_remove(void) {
+    T tte;
+
+    tte.set_future_timestamp(TEST_DELAY_US * 2);
+    int32_t sem_slots = tte.sem_wait(TEST_DELAY_US / 1000);
+    TEST_ASSERT_EQUAL(0, sem_slots);
+    tte.remove();
+
+    sem_slots = tte.sem_wait(TEST_DELAY_US * 2 / 1000 + 1);
+    TEST_ASSERT_EQUAL(0, sem_slots);
+}
+
+/** Template for tests: insert, insert_absolute 0 us timestamp
+ *
+ * Test insert zero
+ * Given an instance of @a TimerEvent subclass
+ * When a timestamp of 0 us is set with @a insert()
+ * Then an event handler is **not** called instantly
+ *     (the event is scheduled after the ticker's overflow)
+ *
+ * Test insert_absolute zero
+ * Given an instance of @a TimerEvent subclass
+ * When a timestamp of 0 us is set with @a insert_absolute()
+ * Then an event handler is called instantly
+ */
+template<typename T>
+void test_insert_zero(void) {
+    T tte;
+
+    tte.set_timestamp(0ULL);
+    int32_t sem_slots = tte.sem_wait(0);
+    TEST_ASSERT_EQUAL(tte.SEM_SLOTS_AFTER_ZERO_TS_INSERTED, sem_slots);
+
+    tte.remove();
+}
+
+/** Template for tests: insert, insert_absolute past timestamp
+ *
+ * Test insert timestamp from the past
+ * Given an instance of @a TimerEvent subclass
+ * When a timestamp of X us is set with @a insert()
+ *     and X is less than present_time
+ * Then an event handler is **not** called instantly
+ *     (the event is scheduled after the ticker's overflow)
+ *
+ * Test insert_absolute timestamp from the past
+ * Given an instance of @a TimerEvent subclass
+ * When a timestamp of X us is set with @a insert_absolute()
+ *     and X is less than present_time
+ * Then an event handler is called instantly
+ */
+template<typename T>
+void test_insert_past(void) {
+    T tte;
+
+    tte.set_past_timestamp();
+    int32_t sem_slots = tte.sem_wait(0);
+    TEST_ASSERT_EQUAL(tte.SEM_SLOTS_AFTER_PAST_TS_INSERTED, sem_slots);
+
+    tte.remove();
+}
+
+utest::v1::status_t test_setup(const size_t number_of_cases)
+        {
+    GREENTEA_SETUP(5, "default_auto");
+    return verbose_test_setup_handler(number_of_cases);
+}
+
+Case cases[] = {
+    Case("Test insert", test_insert<TestTimerEventRelative>),
+    Case("Test insert_absolute", test_insert<TestTimerEventAbsolute>),
+
+    Case("Test remove after insert", test_remove<TestTimerEventRelative>),
+    Case("Test remove after insert_absolute", test_remove<TestTimerEventAbsolute>),
+
+    Case("Test insert zero", test_insert_zero<TestTimerEventRelative>),
+    Case("Test insert_absolute zero", test_insert_zero<TestTimerEventAbsolute>),
+
+    Case("Test insert timestamp from the past", test_insert_past<TestTimerEventRelative>),
+    Case("Test insert_absolute timestamp from the past", test_insert_past<TestTimerEventAbsolute>),
```

Comment:
This test will fail on _NUCLEO_F070RB_ and _EFM32GG_STK3700_ boards until #5051 is fixed.